### PR TITLE
Dynamic provisioning for flocker volume plugin

### PR DESCRIPTION
**Pull request by simonswine**
Refactor flocker volume plugin
- [x] Support provisioning beta (#29006)
- [x] Support deletion
- [x] Use bind mounts instead of /flocker in containers
- [x] support ownership management or SELinux relabeling.
- [x] adds volume specification via datasetUUID (this is guranted to be unique)

I based my refactor work to replicate pretty much GCE-PD behaviour 

**Related issues**: #29006 #26908

@jsafrane @mattbates @wallrj @wallnerryan

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31005)

### CODE REVIEW

File: pkg/volume/flocker/flocker_volume.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+type flockerVolumeProvisioner struct {
+	*flockerVolume
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &flockerVolumeProvisioner{}
+
+func (c *flockerVolumeProvisioner) Provision() (*api.PersistentVolume, error) {
```
Comment (jsafrane) [inline pkg/volume/flocker/flocker_volume.go]: If I am reading it right, you do not support any configuration in StorageClass.Parameters and in PVC.Spec.Selector. In that case, you should check that `c.options.parameters` and `c.options.selector` are empty and throw an error if not.

File: pkg/volume/flocker/flocker_util.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	createOptions := &flockerApi.CreateDatasetOptions{
+		MaximumSize: requestBytes,
+		Metadata: map[string]string{
+			"type": "k8s-dynamic-prov",
+			"pvc":  c.options.PVCName,
+		},
+		Primary: node.UUID,
+	}
+
+	pvName := strings.SplitN(c.options.PVName, "-", 2)
```
Comment (jsafrane) [inline pkg/volume/flocker/flocker_util.go]: I am not sure what is this code supposed to do and especially why. A comment would be helpful.

**Comment by jsafrane** [comment]
I reviewed the provisioner and it looks pretty solid at this point.

File: pkg/volume/flocker/flocker_volume.go (hunk 1 above)
Comment (simonswine) [inline pkg/volume/flocker/flocker_volume.go]: Ok I have added the tests and checks accordingly, with the latest changes

File: pkg/volume/flocker/flocker_util.go (hunk 2 above)
Comment (simonswine) [inline pkg/volume/flocker/flocker_util.go]: I removed the lines. They were supposed to use the PVName UUID for the datasetID in flocker. This seems to be not supported. (Got 400s from Flocker API). So I removed it for now...

**Comment by simonswine** [comment]
@jsafrane Thanks for the review. I've just modified it based on your comments. I think there's no point to hurry, as it won't be merged into 1.4 anyhow?

**Comment by jsafrane** [comment]
> it won't be merged into 1.4 anyhow

No, this is for 1.5. I went through current versions, just a few documentation nits found. Please rebase and finish the deleter!

---

Reviewed 3 of 22 files at r1, 3 of 20 files at r2, 19 of 19 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 746 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-KS6q3bnLpd_6hYomnj_:-KS6q3bo6PPR-BzNy3O5:b-t5h572) ([raw file](https://github.com/kubernetes/kubernetes/blob/fbdbb1b5497d0e5a844e9f198fad3b6823825397/pkg/api/types.go#L746)):*
> ```Go
> // Flocker volumes do not support ownership management or SELinux relabeling.
> type FlockerVolumeSource struct {
> 	// One and only one of the following should be specified.
> ```

IMO, this line should be in FlockerVolumeSource docs, i.e. two lines up ("only one of DatasetName and DatasetUUID should be set"). Otherwise it's godoc of DatasetName, which is quite confusing.

---

*[pkg/api/v1/types.go, line 651 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-KS6qQVTNLa5KiKHLrRo:-KS6qQVUPhmZIOr7hN0t:b-tfnlux) ([raw file](https://github.com/kubernetes/kubernetes/blob/fbdbb1b5497d0e5a844e9f198fad3b6823825397/pkg/api/v1/types.go#L651)):*
> ```Go
> // Flocker volumes do not support ownership management or SELinux relabeling.
> type FlockerVolumeSource struct {
> 	// One and only one of the following should be specified.
> ```

dtto, this is godoc of DatasetName, which does not make sense.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-:-KS6rMnPZ1vuU8FHNaeY:bbaw0an)*

**Comment by simonswine** [comment]
Review status: 0 of 38 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 746 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-KS6q3bnLpd_6hYomnj_:-KSGBJfrAXVnv6sYxVxn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/fbdbb1b5497d0e5a844e9f198fad3b6823825397/pkg/api/types.go#L746)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> IMO, this line should be in FlockerVolumeSource docs, i.e. two lines up ("only one of DatasetName and DatasetUUID should be set"). Otherwise it's godoc of DatasetName, which is quite confusing.
</details>
Done.

---

*[pkg/api/v1/types.go, line 651 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-KS6qQVTNLa5KiKHLrRo:-KSGBNCkZIRk6NUE17If:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/fbdbb1b5497d0e5a844e9f198fad3b6823825397/pkg/api/v1/types.go#L651)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> dtto, this is godoc of DatasetName, which does not make sense.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005)*

**Comment by jsafrane** [comment]
Reviewed 1 of 24 files at r1, 1 of 20 files at r2, 17 of 42 files at r5, 21 of 21 files at r6, 9 of 9 files at r7.
Review status: 28 of 37 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005)*

**Comment by simonswine** [comment]
Review status: 28 of 37 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/volume/flocker/flocker_volume.go, line 55 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005#-KPYX5sSiu7QSZMyzNVm-r1-55:-KSMwEkvKNDxd_UJJE6V:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ae5699c7121dbb7d5f41d7c3b6e84cfb49c3ef3a/pkg/volume/flocker/flocker_volume.go#L55)):*
<details><summary><i>Previously, simonswine (Christian Simon) wrote…</i></summary>
> Ok I have added the tests and checks accordingly, with the latest changes
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005)*

**Comment by simonswine** [comment]
@k8s-bot gci gce e2e test this
@k8s-bot gke e2e test this

**Comment by simonswine** [comment]
@k8s-bot gci gce e2e test this
@k8s-bot gke e2e test this

**Comment by simonswine** [comment]
@jsafrane if you have the chance, could you do another review please? From my point of view it's ready to merge.,

Do you think an exampe is necessary (like the ones in https://github.com/kubernetes/kubernetes/tree/master/examples/experimental/persistent-volume-provisioning)?

**Comment by jsafrane** [comment]
Reviewed 1 of 24 files at r1, 1 of 20 files at r2, 6 of 42 files at r5, 14 of 34 files at r8, 13 of 13 files at r9, 9 of 9 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005)*

**Comment by jsafrane** [comment]
almost LGTM, only all log messages should start with lowercase. Simple grep for glog will reveal them all. I should have noticed it earlier.

**Comment by simonswine** [comment]
@jsafrane Ok I have downcased all the log messages

**Comment by jsafrane** [comment]
Reviewed 1 of 24 files at r1, 1 of 20 files at r2, 6 of 42 files at r5, 5 of 13 files at r9, 1 of 23 files at r11, 13 of 13 files at r12, 9 of 9 files at r13.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31005)*

**Comment by jsafrane** [comment]
LGTM. Thanks a lot for the PR!

**Comment by simonswine** [comment]
@jsafrane I think the only thing missing now is the reviewable "discussions"

**Comment by jsafrane** [comment]
reviewable discussions are optional, important is that "Submit Queue" is green - it's in the queue and it will be merged eventually.

**Comment by simonswine** [comment]
@jsafrane :+1: thanks for clarifying, not too familiar with reviewable yet

**Comment by eddytruyen** [comment]
I get an error from kubelet when mounting flocker volumes using hyperkube:1.5.1
When using hyperkube:1.4.8 everything workes fine.
I do not use dynamic volume provisioning (I.e.I created flocker volume first using flockerctl)

Kube Setup: 
Ubuntu trusty nodes
I am using my own extension to docker-multinode that integrates with flocker: https://github.com/eddytruyen/kube-deploy/tree/master/docker-multinode

Flocker Setup: 1.15
Datasetprovider: Cinder (openstack)

Error log when deploying mysql: https://github.com/kubernetes/kubernetes/tree/master/examples/mysql-wordpress-pd. (I changed the volume YAML file such that a localhost volume is changed into a flocker dataset):

```
  11m           1m              13      {kubelet k8-worker-1}                 
Warning          FailedMount     MountVolume.SetUp failed for volume "kubernetes.io/flocker/38b02b01-dfb5-11e6-83a8-fa163ef52647-local-pv-1" (spec.Name: "local-pv-1") pod "38b02b01-dfb5-11e6-83a8-fa163ef52647" (UID: "38b02b01-dfb5-11e6-83a8-fa163ef52647") with: mount failed: exit status 32
Mounting command: mount
Mounting arguments: /flocker/286d6e19-e715-49e8-9eab-489b325a7d9c /var/lib/kubelet pods/38b02b01-dfb5-11e6-83a8-fa163ef52647/volumes/kubernetes.io~flocker/local-pv-1  [bind]
Output: mount: special device /flocker/286d6e19-e715-49e8-9eab-489b325a7d9c does not exist
```

**Pull request merged**